homeassistant/components/esphomelib/config_flow.py

```diff
@@ -0,0 +1,129 @@
+"""Config flow to configure esphomelib component."""
+import asyncio
+from collections import OrderedDict
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.helpers import ConfigType
+
+
+@config_entries.HANDLERS.register('esphomelib')
+class EsphomelibFlowHandler(config_entries.ConfigFlow):
+    """Handle a esphomelib config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host = None  # type: Optional[str]
+        self._port = None  # type: Optional[int]
+        self._password = None  # type: Optional[str]
+        self._name = None  # type: Optional[str]
+
+    async def async_step_user(self, user_input: Optional[ConfigType] = None,
+                              error: Optional[str] = None):
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            self._host = user_input['host']
+            self._port = user_input['port']
+            error, device_info = await fetch_device_info(self._host,
+                                                         self._port)
+            if error is not None:
+                return await self.async_step_user(error=error)
+            self._name = device_info.name
+
+            # Only show authentication step if device uses password
+            if device_info.uses_password:
+                return await self.async_step_authenticate()
+
+            return self._async_get_entry()
+
+        fields = OrderedDict()
+        fields[vol.Required('host', default=self._host or vol.UNDEFINED)] = str
```

Comment:
What does `vol.UNDEFINED` do here?

Reply:
the `vol.Required` class takes some keyword arguments including `default`. By default it is the sentinel value `vol.UNDEFINED`, so in essence this is the same as

```python
if self._host is not None:
    fields[vol.Required('host')] = str
else:
    fields[vol.Required('host', default=self._host)] = str
```

just shorter. (https://github.com/alecthomas/voluptuous/blob/master/voluptuous/schema_builder.py#L1150)